Mage/MessageStream/MessageStream.cs

```diff
@@ -0,0 +1,202 @@
+			_sessionKey = UnityEngine.WWW.EscapeURL(session["key"].ToString());
+			transportClient.start();
+		});
+		
+		// Stop the message client when session is lost
+		mage.eventManager.on ("session.unset", (object sender, JToken reason) => {
+			transportClient.stop();
+			initializeMessageList();
+			_sessionKey = null;
+		});
+
+		// Set the selected transport client (or the default)
+		this.setTransport (transport);
+	}
+
+
+	// Updates URI and credentials 
+	public void setEndpoint(string baseURL, string username = null, string password = null) {
+		_endpoint = baseURL + "/msgstream";
+		_username = username;
+		_password = password;
+	}
+
+
+	// Sets up given transport client type
+	public void setTransport(TransportType transport) {
+		// Stop existing transport client if any, when nulled out it will be collected
+		// by garbage collecter after existing connections have been terminated.
+		if (transportClient != null) {
+			transportClient.stop ();
+			transportClient = null;
+		}
+
+		// Create new transport client instance
+		if (transport == TransportType.SHORTPOLLING) {
+			Func<string> getShortPollingEndpoint = () => {
+				return getHttpPollingEndpoint ("shortpolling");
+			};
+
+			transportClient = new ShortPolling(getShortPollingEndpoint, getHttpHeaders, processMessagesString, 5000) as TransportClient;
+		} else if (transport == TransportType.LONGPOLLING) {
+			Func<string> getLongPollingEndpoint = () => {
+				return getHttpPollingEndpoint("longpolling");
+			};
+
+			transportClient = new LongPolling(getLongPollingEndpoint, getHttpHeaders, processMessagesString) as TransportClient;
+		} else {
+			throw new Exception("Invalid transport type: " + transport);
+		}
+	}
+
+
+	// Returns the endpoint URL for polling transport clients i.e. longpolling and shortpolling
+	private string getHttpPollingEndpoint(string transport) {
+		string endpoint = _endpoint + "?transport=" + transport + "&sessionKey=" + _sessionKey;
+		if (confirmIds.Count > 0) {
+			endpoint += "&confirmIds=" + string.Join(",", confirmIds.ToArray());
+			confirmIds.Clear();
+		}
+		
+		return endpoint;
+	}
+	
+	
+	// Returns the required HTTP headers
+	private Dictionary<string, string> getHttpHeaders() {
+		if (_username == null && _password == null) {
+			return null;
+		}
+		
+		Dictionary<string, string> headers = new Dictionary<string, string>();
+		string authInfo = _username + ":" + _password;
+		string encodedAuth = Convert.ToBase64String(Encoding.Default.GetBytes(authInfo));
+		headers.Add("Authorization", "Basic " + encodedAuth);
+		return headers;
+	}
+
+
+	// Deserilizes and processes given messagesString
+	private void processMessagesString(string messagesString) {
+		if (messagesString == "" || messagesString == null) {
+			return;
+		}
+
+		try {
+			JObject messages = JObject.Parse(messagesString);
+			addMessages(messages);
+			processMessages();
+		} catch (Exception error) {
+			throw new Exception("Failed to process messages: " + error.ToString());
```

Comment:
yeah i think this was here to give a nice message like "Failed to process messages:" as opposed to some random stack error. Let's discuss whats best way to approach this